core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogInitialization.java

```diff
@@ -528,8 +528,60 @@ private void startBundle(OsgiBundleInstallationResult br) throws BundleException
         }
     }
 
+    private PersistedCatalogState filterPersistedState(PersistedCatalogState persistedState, RebindLogger rebindLogger) {
+        CatalogUpgrades catalogUpgrades = findCatalogUpgrades(rebindLogger);
```

Comment:
feels like this information should be kept in memory not just grabbed for this method.  probably keep in `OsgiManager` or `BrooklynTypeRegistry` ?  not essential for this PR as i expect it will be needed for follow-on work but for cleanliness we should avoid `findCatalogUpgrades` twice

Reply:
Absolutely - intent is to first support these first two options. When we then add support for `brooklyn-catalog-upgrade-for-bundles` and `brooklyn-catalog-upgrade-for-types, the signature of this private method will no doubt change.